ci/pkl-pipelines/common/shared_resources.pkl

```diff
@@ -2,36 +2,129 @@ import "package://pkg.pkl-lang.org/github.com/alphagov/pkl-concourse-pipeline/pk
-        ["branch"] = "master"
+  name = "pay-ci"
+  source = new Mapping {
+    ["uri"] = "https://github.com/alphagov/pay-ci"
+    ["branch"] = "master"
+    ["username"] = "alphagov-pay-ci-concourse"
+    ["password"] = "((github-access-token))"
+  }
+}
+
+function payDockerHubResource(resource_name: String, repo: String, tag: String, checkInterval: String): Pipeline.Resource = new {
+  name = resource_name
+  type = "registry-image"
+  icon = "docker"
+  when (checkInterval.length > 0) {
+    check_every = checkInterval
+  }
+  source = new {
+    ["repository"] = repo
+    ["tag"] = tag
+    ["username"] = "((docker-username))"
+    ["password"] = "((docker-access-token))"
+  }
+}
+
+function payGithubResource(resource_name: String, repo: String, tagRegex: String, path: String, repo_branch: String): Pipeline.Resource = new {
+  name = resource_name
+  type = "git"
+  icon = "github"
+  source = new {
+    ["uri"] = "https://github.com/alphagov/\(repo)"
+    when (tagRegex.length > 0) {
+      ["tag_regex"] = tagRegex
+    }
+    when (repo_branch.length > 0) {
+      ["branch"] = repo_branch
     }
+    ["username"] = "alphagov-pay-ci-concourse"
+    ["password"] = "((github-access-token))"
+    when (path.length > 0) {
+      ["paths"] = new Listing<String> {
+        path
+      }
+    }
+  }
+}
+
+function paySlackNotificationForFail(channel_name: String, message: String): Pipeline.PutStep = new Pipeline.PutStep {
+  put = "slack-notification"
+  attempts = 10
+  params = new Mapping {
+    ["channel"] = channel_name
+    ["silent"] = true
+    ["text"] = ":red-circle: \(message) - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>"
+    ["icon_emoji"] = ":concourse:"
+    ["username"] = "pay-concourse"
+  }
+}
+
+function paySlackNotificationForSuccess(channel_name: String, message: String): Pipeline.PutStep = new Pipeline.PutStep {
+  put = "slack-notification"
+  attempts = 10
+  params = new Mapping {
+    ["channel"] = channel_name
+    ["silent"] = true
+    ["text"] = ":green-circle: \(message) - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>"
+    ["icon_emoji"] = ":concourse:"
+    ["username"] = "pay-concourse"
+  }
 }
 
-function PayDockerHubResource(resource_name: String, repo: String, tag: String): Pipeline.Resource = new {
-    name = resource_name
-    type = "registry-image"
-    icon = "docker"
-    source = new {
-        ["repository"] = repo
-        ["tag"] = tag
-        ["username"] = "((docker-username))"
-        ["password"] = "((docker-access-token))"
+function payECRResource(resource_name: String, repo: String, aws_account_id: String,
+  check_interval: String?, tag: String?, variant: String?): Pipeline.Resource = new {
```

Comment:
I don't think these optional types actually work here, you can't optionally pass arguments, they are always a fixed number of args, that's why you have to call it with empty strings and check for it being empty.

It might be better to leave them off and amend the return value where you want to set it, e.g:
```
(payEcrResource("adminusers-ecr-registry-test", "govuk-pay/adminusers", "01234565")) {
  source { variant = "-release" }
}
```

Or alternatively have a couple of variations of this function, like one for variants, and one for tag?

```
function payECRResource(resource_name: String, repo: String, aws_account_id: String): Pipeline.Resource = new {
  name = resource_name
  type = "registry-image"
  icon = "docker"
  source = new {
    ["repository"] = repo
    ["aws_access_key_id"] = "((readonly_access_key_id))"
    ["aws_secret_access_key"] = "((readonly_secret_access_key))"
    ["aws_session_token"] = "((readonly_session_token))"
    ["aws_role_arn"] = "arn:aws:iam::((\(aws_account_id))):role/concourse"
    ["aws_ecr_registry_id"] = "((\(aws_account_id)))"
    ["aws_region"] = "eu-west-1"
  }
}

function payECRVariantResource(resource_name: String, repo: String, aws_account_id: String, variant: String) = 
(payECRResource(resource_name, repo, aws_account_id)) {
  source {
    ["variant"] = variant
  }
}
```

In this example I also eliminated the check_every, it's used hardly ever in our code, so might be better ammended in the caller on the rare occasion:
```
(payECRVariantResource(...SNIP_ARGS...)) { check_every = "never" }
```

Reply:
Makes sense. Created variant methods for frequently used variables for ECR and Github resources. and amended usages for additional attributes. We can create more variants if required later

Reply:
Awesome